File: 0001~0500/0005. Longest Palindromic Substring/main.cpp
/** 
 * Source: https://leetcode.com/problems/longest-palindromic-substring/solution/
 * Date: 2021/7/22
 * Skill: DP
 * Runtime: 380 ms, faster than 24.58% of C++ online submissions
 * Memory Usage: 29.9 MB, less than 39.09% of C++ online submissions
 * Time complexity:
 * Space complexity: 
 * Constraints: 
 *      1 <= s.length <= 1000
 */

#include<iostream>
#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int sz = s.length(), start = 0, length = 1;
        vector<vector<bool>> dp(sz, vector<bool>(sz+1, true));
        for (int len =  2; len <= sz; len++) {
            for (int i = 0; i+len-1 < sz; i++) {
                if (s[i] == s[i+len-1] && dp[i+1][len-2]) {
                    start = i;
                    length = len;
                } else {
                    dp[i][len] = false;
                }
            }
        }
        return s.substr(start, length);
    }
};


static const auto io_sync_off = []() {
    //* turn off sync
    std::ios::sync_with_stdio(false);
    //* untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0005. Longest Palindromic Substring/main.py
"""
Source: https://rb.gy/mp026
Date: 2023/8/4
Skill:
Ref:
Runtime: 3669 ms, faster than 24.10%
Memory Usage: 25.05 MB, less than 7.25%
Time complexity:
Space complexity:
Constraints:
    1 <= s.length <= 2000
"""

import math
import sys
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


class Solution:
    def longestPalindrome(self, s: str) -> str:
        sz, res = len(s), s[0]
        is_palindrome = [[False for _ in range(sz)] for _ in range(sz)]
        for i in range(sz):
            is_palindrome[i][i] = True
        for i in range(sz - 1):
            if s[i] == s[i + 1]:
                res = s[i: i + 2]
                is_palindrome[i][i + 1] = True
        for length in range(3, sz + 1):
            for i in range(sz - length + 1):
                if s[i] == s[i + length - 1]:
                    is_palindrome[i][i + length - 1] = is_palindrome[i + 1][i + length - 2]
                    if is_palindrome[i][i + length - 1]:
                        res = s[i: i + length]

        return res


if __name__ == "__main__":
    s = Solution()
